db/db.go

```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
+}
+
+func (tx *readTransaction) Close() error {
+	if tx.Closed() {
+		return ClosedError
+	}
+	tx.id = -1
+	removeTransaction(tx)
+	return nil
+}
+
+type writeTransaction struct {
+	id       int
+	db       *DB
+	noms     datas.Database
+	head     Commit
+	me       *kv.MapEditor
+	didWrite bool
+}
+
+func (tx *writeTransaction) ID() int {
+	return tx.id
+}
+
+func (tx *writeTransaction) Closed() bool {
+	return tx.id == -1
+}
+
+func (tx *writeTransaction) Get(id string) ([]byte, error) {
+	if tx.Closed() {
+		return nil, ClosedError
+	}
+	value := tx.me.Get(types.String(id))
+	if value == nil {
+		return nil, nil
+	}
+	var b bytes.Buffer
+	err := nomsjson.ToJSON(value, &b)
+	return b.Bytes(), err
+}
+
+func (tx *writeTransaction) Has(id string) (bool, error) {
+	if tx.Closed() {
+		return false, ClosedError
+	}
+	return tx.me.Has(types.String(id)), nil
+}
+
+func (tx *writeTransaction) Scan(opts ScanOptions) ([]ScanItem, error) {
+	if tx.Closed() {
+		return nil, ClosedError
+	}
+
+	tx.me.Build().NomsMap()
+	// TODO(arv): Implement
+	return nil, nil
+}
+
+func (tx *writeTransaction) Put(path string, JSON []byte) error {
+	if tx.Closed() {
+		return ClosedError
+	}
+
+	canonicalJSON, err := nomsjson.Canonicalize(JSON)
+	if err != nil {
+		return fmt.Errorf("could not Put '%s'='%s': %w", path, JSON, err)
+	}
+	value, err := nomsjson.FromJSON(bytes.NewReader(canonicalJSON), tx.noms)
+	if err != nil {
+		return fmt.Errorf("could not Put '%s'='%s': %w", path, JSON, err)
+	}
+
+	err = tx.me.Set(types.String(path), value)
+	if err != nil {
+		return fmt.Errorf("could not Put '%s'='%s': %w", path, value, err)
+	}
+
+	tx.didWrite = true
+	return nil
+}
+
+func (tx *writeTransaction) Del(path string) (ok bool, err error) {
+	if tx.Closed() {
+		return false, ClosedError
+	}
+
+	k := types.String(path)
+	ok = tx.me.Has(k)
+	if ok {
+		err = tx.me.Remove(k)
+		if err == nil {
+			tx.didWrite = true
+		}
+	}
+	return ok, err
+}
+
+func (tx *writeTransaction) Abort() error {
+	if tx.Closed() {
+		return ClosedError
+	}
+	tx.id = -1
+
+	// No need to do anything.
+	removeTransaction(tx)
+	return nil
+}
+
+func (tx *writeTransaction) Close() error {
+	_, err := tx.Commit()
+	return err
+}
+
+func (tx *writeTransaction) Commit() (ref types.Ref, err error) {
+	if tx.Closed() {
+		err = ClosedError
+		return
+	}
+	tx.id = -1
+
+	removeTransaction(tx)
+	if !tx.didWrite {
+		// No need to do anything.
+		return
+	}
+
+	// Commmit.
+
+	// TOD(arv): name
+	function := "TODO_function_name"
+	// TOD(arv): Args
+	args := types.NewList(tx.noms)
+
+	basis := types.NewRef(tx.head.Original)
+
+	newMap := tx.me.Build()
+	newDataChecksum := newMap.NomsChecksum()
+	newData := tx.noms.WriteValue(newMap.NomsMap())
+
+	commit := makeTx(tx.noms, basis, time.DateTime(), function, args, newData, newDataChecksum)
+
+	ref = tx.noms.WriteValue(commit.Original)
+
+	// FastForward not strictly needed here because we should have already ensured that we were
```

Comment:
This comment is no longer correct. We are now relying on the internal optimistic locking of Noms again.